### PR TITLE
[SPARK-18200][GRAPHX] Support zero as an initial capacity in OpenHashSet

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

[SPARK-18200](https://issues.apache.org/jira/browse/SPARK-18200) reports Apache Spark 2.x raises `java.lang.IllegalArgumentException: requirement failed: Invalid initial capacity` while running `triangleCount`. The root cause is that `VertexSet`, a type alias of `OpenHashSet`, does not allow zero as a initial size. This PR loosens the restriction to allow zero.

## How was this patch tested?

Pass the Jenkins test with a new test case in `OpenHashSetSuite`.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala [hunk 1]
```diff
@@ -42,10 +42,12 @@ import org.apache.spark.annotation.Private
  */
 @Private
 class OpenHashSet[@specialized(Long, Int) T: ClassTag](
-    initialCapacity: Int,
+    var initialCapacity: Int,
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala]: This looks really hacky and makes it a public variable. Why don't we just change the require to >= 0, and then change nextPowerOf2 to pick a value that's greater than 0?

Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala]: Yep. I'll change like that.

File: core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala [hunk 2]
```diff
@@ -271,7 +271,7 @@ class OpenHashSet[@specialized(Long, Int) T: ClassTag](
   private def hashcode(h: Int): Int = Hashing.murmur3_32().hashInt(h).asInt()
 
   private def nextPowerOf2(n: Int): Int = {
-    val highBit = Integer.highestOneBit(n)
+    val highBit = Integer.highestOneBit(if (n == 0) 1 else n)
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala]: can you just do ...

```
if (n == 0) {
  2
} else {
  // original code
}
```

Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala]: Sure.
BTW, for the test failure, I'll look at that soon.

**Comment by dongjoon-hyun** [comment]
The only failure seems to be irrelevant.
```
[info] BasicSchedulerIntegrationSuite:
[info] - job with fetch failure *** FAILED *** (1 second, 889 milliseconds)
[info]   java.util.concurrent.TimeoutException: Futures timed out after [1 second]
```

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by rxin** [comment]
LGTM pending Jenkins.

**Comment by dongjoon-hyun** [comment]
Thank you for review and merging, @rxin .

File: core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala [hunk 3]
```diff
@@ -271,8 +271,12 @@ class OpenHashSet[@specialized(Long, Int) T: ClassTag](
   private def hashcode(h: Int): Int = Hashing.murmur3_32().hashInt(h).asInt()
 
   private def nextPowerOf2(n: Int): Int = {
-    val highBit = Integer.highestOneBit(n)
-    if (highBit == n) n else highBit << 1
+    if (n == 0) {
+      2
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala]: This should have been 1 perhaps ?
nextPowerOf2(1) < nextPowerOf2(0) now ...

Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala]: True or just make the condition <= 1. It won't really matter much but is a little inconsistent.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala]: No I think the issue was that right now nextPowerOf2(1) is 1 and that seems inconsistent if 0 gives 2. Both should result in 2?

Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala]: Then, should I fix this like the following?
```
-      2
+      1
```

**Pull request closed without merging**